### ISSUE TITLE
Should preserve styles of LitElement based web components

### ISSUE THREAD

**Issue by ErikGrimes** [vegarringdal/custom-elements-hmr-polyfill#27]
When using LitElement based web components, the style information in ```static get styles``` is lost.  I'm not sure if there are other issues related to static data.

**Comment by vegarringdal** [comment]
Thanks for reporting this.
I see I have something weird going on in the sample also.
I just havent seen it because in my own code I usually use  I create a new class that extends the elementclass.

Like the `customElementExtended` decortaor here:

https://github.com/vegarringdal/custom-elements-hmr-polyfill/blob/master/src/sample/decorator/defineCustomElement.ts


Ill try and have a look at this during the next days.

**Comment by vegarringdal** [comment]
@LarsDenBakker 

Didnt you have a similar issue we fixed ?
Did something break for you on last update?

**Comment by vegarringdal** [comment]
@ErikGrimes 


I just tested again and looks like it works, maybe I was tricked first time.
Do you have a super simple code sample of code failing ?
Will give me a good start when I get time tomorrow/friday :-)

```ts
import { applyPolyfill, ReflowStrategy } from 'custom-elements-hmr-polyfill';

// add dom content
document.addEventListener('DOMContentLoaded', function(){
    if(document.body.innerHTML.trim() === ''){
        console.log('dom inserted')
        document.body.innerHTML = "<app-one></app-one>"
    
    }
})


// hmr
applyPolyfill(ReflowStrategy.RERENDER_INNER_HTML)


// decorator
export function defineCustomElement(
    elementName: string,
    elementDefinitionOptions?: ElementDefinitionOptions
) {
    return function reg(impl: any) {
        if (elementDefinitionOptions) {
            customElements.define(elementName, impl, elementDefinitionOptions);
        } else {
            customElements.define(elementName, impl);
        }
    };
}


// test element
@defineCustomElement('app-one')
export class ElementX extends HTMLElement {
    static get foo(){
        return 'change me'  // <----- edit me and it gets updated...
    }

    constructor() {
        super();
        console.log('constructor');
    }

    connectedCallback(){
        console.log('connectedCallback');
        this.innerHTML = this.render()
    }

  
    render() {
        console.log('render', ElementX.foo)
        return `
            <h1>${ElementX.foo}</h1>
        `;
    }
}
```

**Comment by vegarringdal** [comment]
@ErikGrimes 

Think I see the issue.

If you want to try a very exprimental buil you can replace dist folder in you apps `node_modules/custom-elements-hmr-polyfil/dist` with this:

https://github.com/vegarringdal/test/tree/master/custom-elements-hmr-polyfill%20-%20next-9

Havent had time to test if I break something else.
Please post a simple sample where you are able to reproduce the bug 😄

**Comment by ErikGrimes** [comment]
Thanks for looking into this so promptly.  Creating a simple sample of the
original issue, may take me a minute.  I tried briefly yesterday, but
wasn't able to reproduce it outside of the original project.  There are a
lot of moving parts (Webpack, Babel, Storybook, etc).  I'm also on vacation
until late next week.

I tried the test build (thanks again!) and the elements in the original
project are being styled now.  This also appears to have fixed another
issue which I hadn't reported yet where styles weren't being applied after
HMR.  Yay! However, the test build breaks Storybook's update of the story
view after an HMR.

*Steps to reproduce*

1. Clone https://github.com/ErikGrimes/storybook-custom-elements-hmr
2. npm install
3. npm run storybook
4. Change src/first-element.js
5. Select the Second Element story in Storybook

*Expected behavior*

Storybook should display SecondElement

*Actual behavior*

Storybook continues to display FirstElement.

**Comment by vegarringdal** [comment]
@ErikGrimes 

* cloned/installed etc
* Updated to the test build I showed you yesterday ( will create a alpha npm package today or tomorrow)
* did some changes to how you use it, since we only want to trigger the root element for rerendering

Personally I prefer to always use `ReflowStrategy.NONE` and be in control.  😄 
Never use delays either since I feel the results gets more laggy, hmr updates something the polyfills edits this again..

```
if (document.body) {
  requestAnimationFrame(() => {
      // re-render the whole DOM
      const root = document.getElementById('root');
      const rootHTML = root.innerHTML;
      root.innerHTML = '';
      root.innerHTML = rootHTML;
  });
}  

applyPolyfill(ReflowStrategy.NONE, null, function(elementName){
  console.log('[HMR] redefine
 element: ', elementName);
}); 
```

Sample of it running (I do not know why the buttons switch locations.. )

![hmr](https://user-images.githubusercontent.com/2901416/67966541-58b77e80-fc04-11e9-9905-8eceb6981b19.gif)

**Comment by vegarringdal** [comment]
@ErikGrimes 
Make a npm dev version `custom-elements-hmr-polyfill@1.0.0-dev.1`

@LarsDenBakker
Can you also try out `custom-elements-hmr-polyfill@1.0.0-dev.1` and see if you get any unexpected behavior?

**Comment by ErikGrimes** [comment]
I made the changes you suggested, used the new dev build in the original,
and at first glance everything looks good!  I'll take a closer look when I
get back from vacation.  As for the switch, it's a weird Storybook thing;
it happens in other projects too :)

**Comment by vegarringdal** [comment]
@ErikGrimes 

Great, Ill wait with any release until you get time to test it so Im sure we dont miss anything.
I need to try and make time to make some tests using cypress.io or something. So I can call it version 1.0.0

**Comment by vegarringdal** [comment]
@ErikGrimes 

Have you had time to test dev version more?
Or still on vacation ? 😄

**Comment by ErikGrimes** [comment]
I am back from vacation, but have been pretty swamped with work.  I added
the polyfill to a large component library with Storybook and had a
colleague go through the catalog.  He did find some issues which we need to
sort through.  I'm hoping we'll get to it later this week or early next
week,

**Comment by vegarringdal** [comment]
@ErikGrimes 

Great.
If you have simple samples where it fails then that will help a lot

**Comment by ErikGrimes** [comment]
@vegarringdal 

We finished sorting through the issues and they appear to be existing issues with our library, not with the polyfill.  So, looks like we're good to go!

**Comment by vegarringdal** [comment]
@ErikGrimes 

Very nice, Ill try and merge my PR during the weekend or early next week.

**Comment by vegarringdal** [comment]
merged and published custom-elements-hmr-polyfill@1.0.0
